### PR TITLE
[FLINK-8069] [table] Support empty watermark strategy for TableSources

### PR DESCRIPTION
**Pull request by xccui**
## What is the purpose of the change

This PR enables an empty watermark strategy for `RowtimeAttributeDescriptor`.

## Brief change log

  - Add a default `null` value for the watermark strategy in `RowtimeAttributeDescriptor`.
  - Add a case in `StreamTableSourceScan` for empty watermark strategy.
  - Add a test case and update related docs.

## Verifying this change

The change can be verified by the new added `testRowtimeTableSourceWithoutWMStrategy()`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: docs/dev/table/sourceSinks.md [hunk 1]
```diff
@@ -316,7 +316,7 @@ A table schema field of type `SQL_TIMESTAMP` can be declared as rowtime attribut
 
 * the name of the field, 
 * a `TimestampExtractor` that computes the actual value for the attribute (usually from one or more other attributes), and
-* a `WatermarkStrategy` that specifies how watermarks are generated for the the rowtime attribute.
+* a `WatermarkStrategy` that specifies how watermarks are generated for the the rowtime attribute (With the default value `null` to indicate no watermark should be generated).
```
Comment (fhueske) [inline docs/dev/table/sourceSinks.md]: I think we should not use `null` to indicate that no watermarks should be assigned. 
Rather provide a `WatermarkStrategy` that indicates that existing watermarks are preserved like:

```
class PreserveWatermarks extends WatermarkStrategy {
  def INSTANCE: PreserveWatermarks = new PreserveWatermarks
}
```

Comment (fhueske) [reply docs/dev/table/sourceSinks.md]: I would leave this part of the docs unchanged

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/definedTimeAttributes.scala [hunk 2]
```diff
@@ -62,12 +62,13 @@ trait DefinedRowtimeAttributes {
   *
   * @param attributeName The name of the rowtime attribute.
   * @param timestampExtractor The timestamp extractor to derive the values of the attribute.
-  * @param watermarkStrategy The watermark strategy associated with the attribute.
+  * @param watermarkStrategy The watermark strategy associated with the attribute. With the default
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/definedTimeAttributes.scala]: revert changes in this class

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/TableSourceITCase.scala [hunk 3]
```diff
@@ -154,6 +154,40 @@ class TableSourceITCase extends StreamingMultipleProgramsTestBase {
   }
 
   @Test
+  def testRowtimeTableSourceWithoutWMStrategy(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/TableSourceITCase.scala]: The test does not really check that watermarks are preserved.

We would need a `TableSource` with a non-parallel `SourceFunction` that emits records and watermarks, similar to `TimeTestUtil.EventTimeSourceFunction`. 

There are two ways to ensure that the watermarks have been preserved:
1. easy / not precise: we emit late data from the TableSource and validated that the late data is not included in the result
2. more complex / precise: We run a simple select project query on the table and convert the query result into an append `DataStream`.  Finally, we need to check that all watermarks have been preserved using a custom operator (other functions do not have access to watermarks).

Comment (fhueske) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/TableSourceITCase.scala]: I'd be fine with either way, but the test should ensure that watermarks are preserved.

**Comment by xccui** [comment]
Thanks for the review @fhueske. I think you are right. A dedicated `WatermarkStrategy` with watermark-preserve semantics  should be better. I'll update the PR according to your comments. Thanks.

**Comment by xccui** [comment]
Hi @fhueske, the PR has been updated. During the refactoring, I got two questions. 

1. Do you think it's still necessary to allow the watermark strategy to be `null`?
2. Currently, we only provide an `ExistingField` extractor for the rowtime field. Considering that the `TableSource` may be created from an existing `DataStream`, I think we should also provide a corresponding extractor, which automatically reads the timestamp from the `StreamRecord` and set it as a field. What do you think?

**Comment by fhueske** [comment]
Hi @xccui, thanks for the update. Regarding your questions:

1. I think `null` should not be supported.
2. Such a `TimestampExtractor` is already available: [`StreamRecordTimestamp`](https://github.com/apache/flink/blob/master/flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/tsextractors/StreamRecordTimestamp.scala) :-)

**Comment by fhueske** [comment]
The PR looks good.
I'll merge it.

**Pull request closed without merging**